sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,446 @@
+
+        Args:
+            theta: Parameters $\theta$.
+            individually: If true, returns log weights and log_probs individually.
+
+        Raises:
+            AssertionError if posterior estimators are a mixture of different methods.
+
+        Returns:
+            `(len(θ),)`-shaped average log posterior probability $\log p(\theta|x)$ for
+            θ in the support of the prior, -∞ (corresponding to 0 probability) outside.
+        """
+        assert all(
+            isinstance(posterior, type(self.posteriors[0]))
+            for posterior in self.posteriors
+        ), "`log_prob()` only works for ensembles of the same type of posterior."
+
+        log_probs = torch.stack(
+            [posterior.log_prob(theta, **kwargs) for posterior in self.posteriors]
+        )
+        log_weights = torch.log(self._weights).reshape(-1, 1)
+
+        if individually:
+            return log_weights, log_probs
+        else:
+            return torch.logsumexp(log_weights.expand_as(log_probs) + log_probs, dim=0)
+
+    def set_default_x(self, x: Tensor) -> "NeuralPosteriorEnsemble":
+        r"""Set new default x for `.sample(), .log_prob()` to use as conditioning context.
+
+        This is a pure convenience to avoid having to repeatedly specify `x` in calls to
+        `.sample()` and `.log_prob()` - only θ needs to be passed.
+
+        This convenience is particularly useful when the posterior ensemble is focused,
+        i.e. has been trained over multiple rounds to be accurate in the vicinity of a
+        particular `x=x_o` (you can check if your posterior object is focused by
+        printing one exemplary component of the ensemble).
+
+        NOTE: this method is chainable, i.e. will return the NeuralPosteriorEnsemble
+        object so that calls like `posterior_enemble.set_default_x(my_x).sample(mytheta)
+        ` are possible.
+
+        Args:
+            x: The default observation to set for every posterior $p_i(theta|x)$ in the
+            ensemble.
+        Returns:
+            `NeuralPosteriorEnsemble` that will use a default `x` when not explicitly
+            passed.
+        """
+        for posterior in self.posteriors:
+            posterior.set_default_x(x)
+        self.potential_fn = EnsemblePotentialProvider(
+            self.posteriors, self._weights, self.prior, x
+        )
+        return self
+
+    def potential(
+        self, theta: Tensor, x: Optional[Tensor] = None, track_gradients: bool = False
+    ) -> Tensor:
+        r"""Evaluates $\theta$ under the potential that is used to sample the posterior.
+        The potential is the unnormalized log-probability of $\theta$ under the
+        posterior.
+        Args:
+            theta: Parameters $\theta$.
+            track_gradients: Whether the returned tensor supports tracking gradients.
+                This can be helpful for e.g. sensitivity analysis, but increases memory
+                consumption.
+        """
```

Comment:
Here, you have to call the [_x_else_default_x](https://github.com/mackelab/sbi/blob/8657c55b39ec8f673a0eaa0cb2428b7f2ce5b14f/sbi/inference/posteriors/base_posterior.py#L134-L145) method. If `x=None`, default_x will be used.